File: tournament.lua
json = require("json")

PAST_OUTCOMES_FILE = BasePath .. "ipc/past_outcomes.json"

MON_DATA_FILE = BasePath .. "mondata/out_json.json"
Mondata = {}
CHEAT_OFFSET = 119

local monDataFile = io.open(MON_DATA_FILE, "r")
if monDataFile then
    local content = monDataFile:read("*all")
    Mondata = json.decode(content)
    monDataFile:close()
else
    console:log("Error: Mondata file not found. Please ensure the file exists at the specified path.")
end

PastOutcomes = {}

function ReadPastOutcomes()
    local file = io.open(PAST_OUTCOMES_FILE, "r")
    if file then
        local content = file:read("*all")
        PastOutcomes = json.decode(content)
        file:close()
    else
        console:log("Error: Past outcomes file not found. Starting with an empty list.")
        PastOutcomes = {}
    end
end

ReadPastOutcomes()


function CurrentMatchup()
    local playerIndexInROM = Mondata[1]["index"]
    local opponentIndexInROM = Mondata[2]["index"]
    local playerIndexInList = 1
    local opponentIndexInList = 2
    local pastOutcomeIndex = 1
    local cheatOffset = CHEAT_OFFSET or 0
    -- console:log("Past outcomes: " .. dump(PastOutcomes))
    while PastOutcomes["outcomes"][tostring(pastOutcomeIndex)] ~= nil do
        local outcome = PastOutcomes["outcomes"][tostring(pastOutcomeIndex)]
        if outcome == 1 then
            -- console:log("Player won against opponent at index " .. pastOutcomeIndex .. ". Moving to next opponent.")
            -- Player won. Move on to the next opponent.
            opponentIndexInList = 2 + pastOutcomeIndex + cheatOffset
            opponentIndexInROM = Mondata[opponentIndexInList]["index"]
        elseif outcome == 2 then
            -- console:log("Player lost against opponent at index " .. pastOutcomeIndex .. ". Switching player and opponent.")
            -- Player lost. Opponent becomes the next player, and the next opponent is chosen.
            playerIndexInROM = opponentIndexInROM
            playerIndexInList = opponentIndexInList
            opponentIndexInList = 2 + pastOutcomeIndex + cheatOffset
            opponentIndexInROM = Mondata[opponentIndexInList]["index"]
        else
            console:log("Invalid outcome value: " .. tostring(outcome) .. ". Expected 1 (win) or 2 (loss).")
        end

        pastOutcomeIndex = pastOutcomeIndex + 1
    end
    playerName = Mondata[playerIndexInList]["trainer_class"] .. " " .. Mondata[playerIndexInList]["character_name"]
    opponentName = Mondata[opponentIndexInList]["trainer_class"] .. " " .. Mondata[opponentIndexInList]["character_name"]
    console:log(string.format("Battle %d", pastOutcomeIndex))
    console:log(string.format("Current matchup: Player: %s (Index in ROM: %d) vs Opponent: %s (Index in ROM: %d)", playerName, playerIndexInROM, opponentName, opponentIndexInROM))
    return {
        player = playerIndexInROM,
        opponent = opponentIndexInROM,
        index = pastOutcomeIndex
    }
end

function EndCurrentMatchup(outcome)
    local currentMatchup = CurrentMatchup()
    local pastOutcomeIndex = currentMatchup["index"]
    PastOutcomes["outcomes"][tostring(pastOutcomeIndex)] = outcome

    -- Save the updated outcomes back to the file
    local file = io.open(PAST_OUTCOMES_FILE, "w")
    if file then
        local content = json.encode(PastOutcomes)
        file:write(content)
        file:close()
    else
        console:log("Error: Could not open past outcomes file for writing.")
    end
end

File: tournament_py/tournament.py
#!/usr/bin/python

import json
from dataclasses import dataclass, field
from tournament_py.trainer_class_names import trainer_class_names
import string

MON_DATA_FILE = "mondata/out_json.json"
PAST_OUTCOMES_FILE = "ipc/past_outcomes.json"

@dataclass
class TournamentData:
    outcomes_per_event: dict[int, int] = field(default_factory=dict)
    current_player_json: dict = field(default_factory=dict)
    current_opponent_json: dict = field(default_factory=dict)
    last_player_json: dict = field(default_factory=dict)
    last_opponent_json: dict = field(default_factory=dict)
    last_winner_json: dict = field(default_factory=dict)
    last_loser_json: dict = field(default_factory=dict)
    streak_by_index: dict[int, int] = field(default_factory=dict)
    current_streak: int = 0

mon_data = {}
with open(MON_DATA_FILE, 'r') as f:
    mon_data = json.load(f)

def ReadPastOutcomes():
    past_outcomes = {}
    with open(PAST_OUTCOMES_FILE, 'r') as f:
        past_outcomes = json.load(f)
    return past_outcomes

def TrainerData(index):
    global mon_data
    return mon_data[index]

def ParseTournamentData(past_outcomes):
    playerIndexInList = 0
    opponentIndexInList = 1
    pastOutcomeIndex = 1
    result = TournamentData()
    lastPlayerIndexInList = 0
    lastOpponentIndexInList = 1
    lastWinnerIndex = 0
    lastLoserIndex = 0
    while str(pastOutcomeIndex) in past_outcomes["outcomes"]:
        lastPlayerIndexInList = playerIndexInList
        lastOpponentIndexInList = opponentIndexInList
        outcome = past_outcomes["outcomes"][str(pastOutcomeIndex)]
        result.outcomes_per_event[pastOutcomeIndex] = outcome
        if outcome == 1:
            # -- Player won. Move on to the next opponent.
            # print(TrainerData(playerIndexInList)["name"] + " won against opponent " + TrainerData(opponentIndexInList)["name"] + " in event " + str(pastOutcomeIndex))
            lastWinnerIndex = playerIndexInList
            lastLoserIndex = opponentIndexInList
            result.streak_by_index[playerIndexInList] = result.streak_by_index.get(playerIndexInList, 0) + 1
            opponentIndexInList = 1 + pastOutcomeIndex
        elif outcome == 2:
            # -- Player lost. Opponent becomes the next player, and the next opponent is chosen.
            # print(TrainerData(opponentIndexInList)["name"] + " won against player " + TrainerData(playerIndexInList)["name"] + " in event " + str(pastOutcomeIndex))
            lastWinnerIndex = opponentIndexInList
            lastLoserIndex = playerIndexInList
            result.streak_by_index[opponentIndexInList] = result.streak_by_index.get(opponentIndexInList, 0) + 1
            playerIndexInList = opponentIndexInList
            opponentIndexInList = 1 + pastOutcomeIndex
        else:
            print("Invalid outcome for event " + str(pastOutcomeIndex) + ": " + str(outcome))

        pastOutcomeIndex += 1
    result.current_player_json = TrainerData(playerIndexInList)
    result.current_opponent_json = TrainerData(opponentIndexInList)
    result.last_player_json = TrainerData(lastPlayerIndexInList)
    result.last_opponent_json = TrainerData(lastOpponentIndexInList)
    result.last_winner_json = TrainerData(lastWinnerIndex)
    result.last_loser_json = TrainerData(lastLoserIndex)
    result.current_streak = result.streak_by_index.get(playerIndexInList, 0)
    return result

def TrainerFullName(trainer_json):
    class_id = trainer_json["trainer_class"]
    class_name = trainer_class_names.get(class_id, "Pokemon Trainer")
    return string.capwords(f"{class_name} {trainer_json['character_name']}")